### ISSUE TITLE
feature: trace group-level standard deviations in mixed-effects models

### ISSUE THREAD

**Issue by schmettow** [stan-dev/rstanarm#112]
#### Summary:

MCMC samples of should include the standard deviation of random effects as this is a parameter of interest (and not nuisance)
#### Description:

Currently, rstanarm only saves the individual means of random effects in the posterior. While this is of use in some applications (e.g. psychometric), more commonly random effects are used to describe group-level variation. It is possible to reconstruct these parameters from the individual levels. However, this is rather unhandy for researchers (like me), who routinely examine and compare random effect variation. The other two Bayesian  glmm packages brms and MCMCglmm do this. The latter even has as a default to only save the group-level SD.

**Comment by bgoodri** [comment]
I don't understand. If you just do
```{R}
library(rstanarm)
example(example_model, run.dontrun = TRUE)
example_model$ses
```
it has the standard deviation of the herd intercepts.

**Comment by jgabry** [comment]
Yeah those SDs are definitely returned. But my hunch is that the intent was to
ask about the hierarchical variance/sd parameters, in which case the
language was confusing but it's true we don't store them.

**Comment by schmettow** [comment]
This is, indeed, what I meant: for my analysis i need the full posterior of hierarchical SD/variance parameters, not just the MAPs. The same actually goes for corr/cov between random effects. Sorry for the confusing language.

Btw., i don't need any high level functions, as I usually extract the posterior myself and work with my own summary routines. So, if rstanarm:::as.data.frame.stanreg(model) would just include the additional parameters, i'd be perfectly happy.

**Comment by jgabry** [comment]
@bgoodri it's been a while since we revisited the idea of including the
draws of these parameters. Do you still think it's not worth it to
construct them? We would have to save all the thetaL parameters and then
feed them through VarCorr one draw at a time right?

**Comment by bgoodri** [comment]
It's rare that someone would want more than what VarCorr produces now.

**Comment by jgabry** [comment]
On Fri, Aug 26, 2016 at 1:22 AM, bgoodri <notifications@github.com> wrote:

> It's rare that someone would want more than what VarCorr produces now.
>

@bgoodri I think we've received several questions on stan-users about this
and now this github issue, and I was asked about this when I did the short
course, and I'm sure there are more people with this question who just
didn't take the time to write to us. If we're not going to include them
then I would prefer to at least document our reasoning somewhere, because
from a user's perspective I don't think it's clear why draws for some
parameters can be accessed and others can't.

Do you think those draws aren't useful for anything or just that it's not
worth the effort to write the code to construct them? If the former, can
you explain why you think people shouldn't care about those draws? I can
add something to the doc, but it's not clear to me why people shouldn't
want to look at more than point estimates of those parameters.

**Comment by schmettow** [comment]
maybe a concrete example helps: in the attached paper we compare four random effects in their relative contribution to performance. As you can see, the full posteriors were plotted and that is quite informative. The analysis was implemented in Stan. So, I wonder why it is such a big deal to just save these parameters with the posterior. I am currently working on a book for Springer "New statistics for the design researcher", where I would love to use rstanarm all the way through, but this is a showstopper.
[Egans conjecture final.pdf](https://github.com/stan-dev/rstanarm/files/456366/Egans.conjecture.final.pdf)

**Comment by jgabry** [comment]
Thanks for following up. I'm not opposed to this and I certainly don't want
to prevent you from using rstanarm in your book!

Because of the way we parameterize the models and how the code is written I
think it's not quite as simple as just not discarding the draws for those
parameters, but I don't think it would be particularly difficult to
do.  @bgoodri wrote that part of the code so he would know best.

Ben, I know that for many applications the full posterior for these
parameters isn't essential, but given this issue and the questions we've
gotten (on Stan-users and in person -- see my previous response on this
thread), would you be open to adding this for one of the next rstanarm
releases?

**Comment by bgoodri** [comment]
It would probably be easier to add an `anova.stanreg.merMod` method to rstanarm that returns the posterior distribution of the group level (co)variance terms than it would be to store that in the Stan program. Even if we stored the stuff from Stan, we would have to map it through `lme4::VarCorr` in R to get the inputs to the plots in your paper.

Since lme4 has an `anova.merMod` method, I guess that rstanarm should have a Bayesian one.  But I don't think anyone should call it. When someone claims that

> differences among people usually account for much more variability __in performance__ than differences in system designs

then there are several ways to go about assessing it that do not involve plots of posterior standard deviations of group-level terms the contribute to the linear predictor of a Poisson model with an exponential inverse link function. 

One is to estimate a few models and compare them with `loo` to see which is expected to predict new data best. For example, a model with deviations in the intercept for each system design or a model with both person-level deviations and system-level deviations is expected to predict better than a model with person-level deviations alone.

Or you could estimate a model with person-level deviations, system-level deviations, and their interaction, but when you call `posterior_predict`, do it first with `re.form = NULL`, then with `re.form = ~ (1 | person)`, and then with `re.form = ~ (1 | system)`. This allows you to compare the variance __in the predicted performance__ in these three scenarios rather than comparing `sigma_person` to `sigma_system`. Whether a difference between `sigma_person` and `sigma_system` is substantively meaningful depends on the rest of the model in ways that are not evident from a plot of their marginal distributions. More generally, we discourage people from interpreting the posterior distribution of parameters in a (nonlinear) model and encourage people to interpret predictive distributions. 

TL; DR version: I'll put it in (probably not this week but in the next rstanarm after that) but I really hate it.

**Comment by jgabry** [comment]
On Tuesday, September 6, 2016, bgoodri <notifications@github.com> wrote:

> It would probably be easier to add an anova.stanreg.merMod method to
> rstanarm that returns the posterior distribution of the group level
> (co)variance terms than it would be to store that in the Stan program. Even
> if we stored the stuff from Stan, we would have to map it through
> lme4::VarCorr in R to get the inputs to the plots in your paper. Since
> lme4 has an anova.merMod method, I guess that rstanarm should have a
> Bayesian one.  But I don't think anyone should call it.
>
Ok thanks Ben. I think we should add this but also put in the documentation
the reasons you give below for not using it. At least that way when we get
questions about it we can point to the doc. If you add the function I can
add the doc and I can also add a section to one of the vignettes
demonstrating the preferable methods you outline below.

> More generally, we discourage people from interpreting the posterior
> distribution of parameters in a (nonlinear) model and encourage people to
> interpret predictive distributions.
>
Totally agree.

> TL; DR version: I'll put it in (probably not this week but in the next
> rstanarm after that) but I really hate it.
>
Ok thanks. Right now it's mysterious to most users why we don't have this,
so even if it's discouraged I think adding it and discouraging it
(demonstrating preferred methods) is the best way to go.

@schmettow Would using the proposed new function or one of @bgoodri's
preferred methods be sufficient for your book?

**Comment by schmettow** [comment]
Hi jgraby,

as long as this function contains the (virtual) posterior samples in the order of occurrence, i would be fine. No problem to wrap it for my own purposes. Still, I am wondering, how do other users of rstanarm check convergence of the variance parameters, caterpillar plots and the like. This leaves me a bit uncomfortable.

Furthermore, I was thinking about bgoodri's worries regarding my approach: am I completely wrong in taking the group-level sigma as a measure for the amount of variation (up to ranks)?

Thanks again,
Martin.

**Comment by jgabry** [comment]
Hi Martin, some comments below:

On Monday, September 12, 2016, schmettow <notifications@github.com> wrote:

> Hi jgraby,
>
> as long as this function contains the (virtual) posterior samples in the
> order of occurrence, i would be fine. No problem to wrap it for my own
> purposes. Still, I am wondering, how do other users of rstanarm check
> convergence of the variance parameters, caterpillar plots and the like.
> This leaves me a bit uncomfortable.
>
For the particular models that rstanarm implements, there shouldn't
be convergence problems that can only be found by looking directly at those
parameters. But I do think we should allow the user to access them, in
which case you could check them if you want.


> Furthermore, I was thinking about bgoodri's worries regarding my approach:
> am I completely wrong in taking the group-level sigma as a measure for the
> amount of variation (up to ranks)?
>
Not completely wrong, but I think @bgoodri's point was that it's much more
informative about the outcome variable of interest to instead focus on the
differing variabilities in the posterior predictive distribution. It's hard
to get anything super useful from the group-level standard deviation
parameters with the nonlinear transformation required for the glm.  Instead
you want to look at variation in the implied predictive distribution of the
outcome which is much easier to interpret

> Thanks again,
> Martin.
>
My pleasure. Thanks for the feedback on rstanarm. These are definitely
issues worth discussing.

Jonah

>

**Comment by davharris** [comment]
>It's hard to get anything super useful from the group-level standard deviation parameters with the nonlinear transformation required for the glm

I disagree; one can compare the magnitude to the magnitude of other pre-transformation quantities.

Also, not all rstanarm models include a nonlinear transform. Sometimes the response variable is Gaussian.

It does seem strange to me from a Bayesian perspective that rstanarm provides point estimates for quantities like these but not the full posterior.

**Comment by jgabry** [comment]
Hi David, yes I agree that sometimes it can be useful to directly compare
the group-level sds, but I agree with Ben that it's typically less useful
than most applied researchers realize. That is, we see far too often
that people focus on parameters when what they claim to be doing is
focusing on the performance of the model (I'm not implicitly including you
among those people). So we try to encourage the posterior predictive
distribution. That said, I do agree with you that it's strange to not at
least provide access to the posterior for those parameters, even if we
(often) recommend a different approach.

Jonah

**Comment by jgabry** [comment]
Can close this finally! Hierarchical SD/variance parameters are now available on the master branch and in the rstanarm just submitted to CRAN (binaries won't be available from CRAN for a few days though probably).

Here's an example: 
```r
fit <- stan_glmer(mpg ~ wt + (1 + wt|cyl), data = mtcars)
draws <- as.matrix(fit)
colnames(draws)

 [1] "(Intercept)"                    "wt"                            
 [3] "b[(Intercept) cyl:4]"           "b[wt cyl:4]"                   
 [5] "b[(Intercept) cyl:6]"           "b[wt cyl:6]"                   
 [7] "b[(Intercept) cyl:8]"           "b[wt cyl:8]"                   
 [9] "sigma"                          "Sigma[(Intercept),(Intercept)]"
[11] "Sigma[wt,(Intercept)]"          "Sigma[wt,wt]"  
``` 
The last three parameters are the three unique parameters in the covariance matrix. There's no `Sigma[(Intercept),wt]` because it has the same value as `Sigma[wt,(Intercept)]`.
